comparison/loadtest.go

```diff
@@ -216,12 +223,43 @@ func initLoadTest(t *terraform.Terraform, buildCfg BuildConfig, dumpFilename str
 		return fmt.Errorf("invalid db engine %s", dpConfig.TerraformDBSettings.InstanceEngine)
 	}
 
+	resetBucketCmds := []localCmd{}
+	if s3BucketURI != "" && tfOutput.HasS3Bucket() {
+		deleteBucketCmd := localCmd{
+			msg:   "Emptying S3 bucket",
+			value: []string{"--profile", t.Config().AWSProfile, "s3", "rm", "s3://" + tfOutput.S3Bucket.Id, "--recursive"},
+		}
+
+		prepopulateBucketCmd := localCmd{
+			msg:   "Pre-populating S3 bucket",
+			value: []string{"--profile", t.Config().AWSProfile, "s3", "cp", s3BucketURI, "s3://" + tfOutput.S3Bucket.Id, "--recursive"},
+		}
+
+		resetBucketCmds = []localCmd{deleteBucketCmd, prepopulateBucketCmd}
+	}
+
 	if dumpFilename == "" {
 		cmds = append(cmds, startCmd, createAdminCmd, initDataCmd)
 	} else {
 		cmds = append(cmds, loadDBDumpCmd, startCmd)
 	}
 
+	// Resetting the buckets can happen concurrently with the rest of the remote commands,
+	// but we need to cancel them on return in case we return early (as when we receive from cancelCh)
+	resetBucketCtx, resetBucketCancel := context.WithCancel(context.Background())
+	defer resetBucketCancel()
+	resetBucketErrCh := make(chan error)
+	go func() {
+		for _, c := range resetBucketCmds {
+			mlog.Info(c.msg)
+			if err := exec.CommandContext(resetBucketCtx, "aws", c.value...).Run(); err != nil {
+				resetBucketErrCh <- fmt.Errorf("failed to run local cmd %q: %w", c.value, err)
+				return
+			}
+		}
+		resetBucketErrCh <- nil
```

Comment:
Wouldn't it be better to close the channel here, that way we release the resources directly and readers still get `nil` when reading.

Reply:
I feel like sending `nil` explicitly is more semantic, but I'm 2/5. In [A Tour of Go](https://go.dev/tour/concurrency/4), it says:
> Channels aren't like files; you don't usually need to close them. Closing is only necessary when the receiver must be told there are no more values coming, such as to terminate a range loop. 

If you have a stronger opinion, I can change it. Thoughts?

Reply:
The 'need' does a lot of heavy lifting here. They mean that you wouldn't leak resources if you don't close it. I'm 2/5 as well, it doesn't really matter in this case.